File: jupyter/lamp.py
import os
import re
import csv
import json
import itertools
import urllib.request as request

# The root type in LAMP. You must use LAMP.connect(...) to begin using any LAMP classes.
class LAMP(object):
    
    # Basic data model methods that all derived types inherit.
    def __init__(self):
        pass
    def __init__(self, json_dict):
        for k, v in json_dict.items():
            self.__dict__[k] = v
    def __str__(self):
        return {**{ '_type': type(self) }, **self.__dict__}.__str__()
    def __repr__(self):
        return {**{ '_type': type(self) }, **self.__dict__}.__repr__()
    def __hash__(self):
        return hash(self.__dict__['id']) if 'id' in self.__dict__ else 0
    def __eq__(self, other):
        if isinstance(other, self.__class__) and ('id' in self.__dict__ and 'id' in other.__dict__):
            return self.__dict__['id'] == other.__dict__['id']
        else:
            return False
    def __ne__(self, other):
        return not self.__eq__(other)
    
    # The headers for the client, as sent by the LAMP API server.
    header = { }
    
    # Registry of all generated remote types available on the LAMP API server.
    # The base type for all of these generated types, LAMP, is also present.
    types = { }
    
    # Connect to an instance of a server hosting the LAMP API and mirror it locally.
    @staticmethod
    def connect(root_url, auth_type, username, password):
        
        # Download the API definition from the server at root_url.
        api = json.loads(request.urlopen(root_url).read())
        LAMP.header = api.pop('LAMP', None)
        LAMP.types['LAMP'] = LAMP
        
        # Invoke the REST endpoint call and produce local LAMP objects. [Runtime]
        def get_rest(sys, url):
            
            # Get the JSON from the remote server and convert it into LAMP objects. [Runtime]
            # This is invoked at runtime via `LAMP.Study.view('...')`, etc.
            def runtime_json(*x):
                x = list(x)
                if x[0] is ():
                    x.pop(0)
                
                # Parse the base url and replace URL components. [Runtime]
                base = url.replace('GET ', '')
                if base.count('@') is not len(x):
                    raise ValueError('incorrect number of arguments for this endpoint (expected {0}, received {1})'.format(base.count("@"), len(x)))
                base = re.sub("(@[a-zA-Z0-9_]+)", lambda m: x.pop(0)[0], base)
                
                # Obtain JSON and throw an error if that was the case. [Runtime]
                data = {}
                try:
                    auth = '?auth=' + ':'.join([auth_type, username, password])
                    data = request.urlopen(root_url + base + auth).read()
                except urllib.error.HTTPError as error:
                    data = error.read()
                data = json.loads(data)
                if ('error' in data) or ('result' not in data):
                    raise ValueError(data['error']) if 'error' in data else ValueError('no data returned')
                
                # If not error, map each result to an object. [Runtime]
                out = list(map(lambda r: LAMP.types[sys](r), data['result']))
                return out[0] if len(out) == 1 else out
            
            # The attachment to a LAMP derived class is a static lambda calling the above^.
            return staticmethod(lambda *x: runtime_json(x))
        
        # Parse all remote declared types and transform them into local classes.
        for sys, obj in api.items():
            
            # Convert properties to None for now, and endpoints to static methods.
            props = {x: None for x in obj.get('properties', {}).keys()}
            ends = {name: get_rest(sys, url) for name, url in obj.get('endpoints', {}).items()}
            t = type(sys, (LAMP,), {**props, **ends})
            
            # Make the new class accessible (and indexed) under LAMP as static properties.
            setattr(LAMP, sys, t)
            LAMP.types[sys] = t

File: src/lamp.js
/* eslint-disable */

// The root type in LAMP. You must use `LAMP.connect(...)` to begin using any LAMP classes.
export default class LAMP {

    // ActivitySpecID for anything that is a survey
    static get SURVEY_SPEC() {
        return 'QWN0aXZpdHlTcGVjOjEK';
    }

    // ActivitySpecID for anything that is a batch
    static get BATCH_SPEC() {
        return 'QWN0aXZpdHlTcGVjOjAK';
    }

    // Connect to an instance of a server hosting the LAMP API and mirror it locally.
    static async connect(root_url) {

        // Convenience method to wrap most of the Fetch API's nuances away.
        async function _fetch(method, base, route, data) {
            const options = {
                method: method,
                headers: new Headers({
                    'Authorization': `Basic ${LAMP.auth.id}:${LAMP.auth.password}`,
                })
            }
            if (data !== undefined)
                options.body = JSON.stringify(data)

            var res = await (await fetch(`${base}${route}`, options)).json()

            if (res['fatal'] !== undefined)
                throw new Error('SERVER: ' + res['fatal']['message'] + '\n' + res['fatal']['trace'])
            else if (res['error'] !== undefined)
                throw new Error(res['error'])
            return res
        }

        // Invoke the REST endpoint call and produce local LAMP objects. [Runtime]
        // Get the JSON from the remote server and convert it into LAMP objects. [Runtime]
        // This is invoked at runtime via `LAMP.Study.view('...')`, etc.
        async function _get_rest(sys, http_method, base, args) {
            // Parse the base url and replace URL components. [Runtime]
            let frags = base.replace(/({[a-zA-Z0-9_]+})/g, (x) => {
                let y = args.shift()
                if (y === undefined)
                    throw new Error('expected more arguments than provided')
                return y
            })

            // Get request body and handling hints if provided.
            let body = args.shift()
            let hints = args.shift()
            if (args.length > 0)
                throw new Error('expected fewer arguments than provided')

            // Obtain JSON and throw an error if that was the case. [Runtime]
            let data = await _fetch(http_method, root_url, frags, body)
            if (data.data === undefined)
                throw new Error('no data returned')

            // If the API hints includes `untyped`, don't typecast it!
            if (hints !== undefined && hints.untyped === true)
                return data.data

            // If not error, map each result to an object. [Runtime]
            return data.data.map(x => Object.assign(new LAMP.types[sys](), x))
        }

        // Download the API definition from the server at root_url.
        LAMP.auth = {type: null, id: null, password: null}
        let api = await (await fetch(`${root_url}`, {method: 'GET'})).json()
        LAMP.typedef = api

        // Registry of all generated remote types available on the LAMP API server.
        // The base type for all of these generated types, LAMP, is also present.
        LAMP.types = { "LAMP": LAMP }

        // Parse all remote declared types and transform them into local classes.
        for (let sys in api.components.schemas) {
            let obj = api.components.schemas[sys]
            let t = eval(`(function() { function ${sys} () {}; ${sys}.prototype = Object.create(LAMP.prototype); return ${sys}; })()`)

            // Convert properties to null keys and endpoints to static methods. 
            Object.entries(obj.properties || {}).forEach(prop => {
                t.prototype[prop] = null
            })
            Object.entries(api.paths)
                .filter(x => Object.values(x[1])[0].operationId.startsWith(sys + '::'))
                .forEach(endpoint => {                    
                    Object.entries(endpoint[1]).forEach(x => {
                        var method = x[0].toUpperCase()
                        t[x[1].operationId.split('::')[1]] = async function() {
                            return await _get_rest(sys, method, endpoint[0], Array.from(arguments))
                        }
                    })
                })

            // Make the new class accessible under LAMP as static properties.
            LAMP[sys] = t
            LAMP.types[sys] = t
        }

        // Load a session-stored authorization token if available.
        let _saved = JSON.parse(sessionStorage.getItem('LAMP.auth')) || LAMP.auth
        await LAMP.set_identity(_saved.type, _saved.id, _saved.password)
    }

    // Authenticate/authorize as a user of a given `type`.
    // If all values are null (especially `type`), the authorization is cleared.
    static async set_identity(type = null, id = null, password = null) {
        let l = LAMP.auth || {type: null, id: null, password: null}
        if (l.type === type && l.id === id && l.password === password)
            return

        try {
            LAMP.auth = {type: type, id: id, password: password}
            if (type === 'root') {
                LAMP.me = await LAMP.Researcher.all()
            } else if (type === 'researcher') {
                LAMP.me = (await LAMP.Researcher.view('me'))[0]
            } else if (type === 'participant') {
                LAMP.me = (await LAMP.Participant.view('me'))[0]
            } else {
                LAMP.me = null
            }
        } catch(err) {
            LAMP.auth = {type: null, id: null, password: null}
            LAMP.me = null
            throw err
        } finally {
            if (LAMP.auth.type === null && LAMP.auth.id === null && LAMP.auth.password === null)
                LAMP.auth = null
            sessionStorage.setItem('LAMP.auth', JSON.stringify(LAMP.auth))
        }
    }

    // Get the current authenticated/authorized identity.
    static get_identity() {
        return LAMP.me
    }
}
